src/sql/src/plan/statement/ddl.rs

```diff
@@ -922,28 +796,15 @@ pub fn plan_create_source(
             }
             SourceEnvelope::Debezium(dedup_strat, mode)
         }
-        sql_parser::ast::Envelope::Upsert(key_format) => match connector {
-            Connector::Kafka { .. } => {
-                let mut key_encoding = if key_format.is_some() {
-                    get_encoding(key_format)?
-                } else {
-                    encoding.clone()
-                };
-                match &mut key_encoding {
-                    DataEncoding::Avro(AvroEncoding {
-                        key_schema,
-                        value_schema,
-                        ..
-                    }) => {
-                        if key_schema.is_some() {
-                            *value_schema = key_schema.take().unwrap();
-                        }
-                    }
-                    DataEncoding::Bytes | DataEncoding::Text => {}
-                    _ => unsupported!("format for upsert key"),
+        sql_parser::ast::Envelope::Upsert => match connector {
+            Connector::Kafka { .. } => match format {
+                CreateSourceFormat::KeyValue { .. } => SourceEnvelope::Upsert,
+                // this is validated to become keyvalue in purify, where we can talk to CSR
```

Comment:
I'm not sure I follow this comment.

Reply:
I think the point is that logically upsert sources need both a key and value format, which is why it's confusing to accept `Bare` here.

However, to match historical practice, we allow _only_ a value format if and only if it happens to be Avro + CSR. In that case, we then get the corresponding key format from CSR as well.

(This could indeed use some expansion to be comprehensible to people who don't understand the historical context).

Also now I can't find the code that does what I'm talking about so maybe I'm just totally confused and @quodlibetor can correct me.

Reply:
oh, you're correct brennan but the comment is slightly wrong: it still happens in two parts:

* We determine that there is a key/value *schema* in purify
* We convert from a single term to a keyvalue term when we rewrite from `Format` -> `Encoding` later

There is a small comment on [line 949 of `plan/statement/ddl.rs`](https://github.com/MaterializeInc/materialize/pull/6487/files#diff-c840a7fbb6a642d6cbf71dddd92ec1b6dfa86310022ead5673e8fa509ba73dc4R949) (inside `get_encoding`) that mentions the fact that this is ... bad. IMO it _should_ be in purify, but as mentioned this PR is already too large.